contracts/vault/src/execute.rs

```diff
@@ -0,0 +1,591 @@
+fn execute_deposit(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    amount: Uint128,
+) -> StdResult<Response> {
+    check_not_paused(&deps.as_ref())?; // Ensure the contract is not paused
+    let config = state::CONFIG.load(deps.storage)?; // Load configuration
+    let sender = info.sender.clone(); // Address of the user depositing
+
+    // Verify that USDC was sent and matches the specified amount
+    let usdc = info
+        .funds
+        .iter()
+        .find(|c| c.denom == config.usdc_denom)
+        .ok_or_else(|| StdError::generic_err("No USDC sent"))?;
+    if usdc.amount != amount {
+        return Err(StdError::generic_err("Mismatched USDC amount"));
+    }
+
+    // Calculate the amount of USDCLP tokens to mint based on total assets and LP supply
+    let total_assets = get_total_assets(deps.as_ref(), &env)?;
+    let total_lp = state::TOTAL_LP_SUPPLY.load(deps.storage)?;
+    let lp_amount = if total_lp.is_zero() {
+        amount
+    } else {
+        amount.multiply_ratio(total_lp, total_assets)
+    };
+
+    // Create a message to mint USDCLP tokens
+    let mint_msg = WasmMsg::Execute {
+        contract_addr: config.usdclp_address.to_string(),
+        msg: to_json_binary(&Cw20ExecuteMsg::Mint {
+            recipient: sender.to_string(),
+            amount: lp_amount,
+        })?,
+        funds: vec![],
+    };
+
+    // Update the total LP token supply
+    state::TOTAL_LP_SUPPLY.update(deps.storage, |t| -> Result<Uint128, StdError> {
+        Ok(t + lp_amount)
+    })?;
+
+    // Return a response with the mint message and attributes
+    Ok(Response::new().add_message(mint_msg).add_attributes(vec![
+        ("action", "deposit"),
+        ("user", sender.as_str()),
+        ("amount", &amount.to_string()),
+        ("lp_minted", &lp_amount.to_string()),
+    ]))
+}
+
+/// Registers a withdrawal request by burning USDCLP tokens
+///
+/// # Parameters
+/// - `deps`: Mutable dependencies
+/// - `_env`: Contract environment (unused)
+/// - `info`: Message information
+/// - `amount`: Amount of USDCLP to burn
+///
+/// # Returns
+/// - `StdResult<Response>`: Response with burn message and attributes
+fn execute_request_withdrawal(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    amount: Uint128,
+) -> StdResult<Response> {
+    check_not_paused(&deps.as_ref())?; // Ensure the contract is not paused
+    let config = state::CONFIG.load(deps.storage)?; // Load configuration
+    let sender = info.sender.clone(); // Address of the user requesting withdrawal
+
+    // Create a message to burn the user's USDCLP tokens
+    let burn_msg = WasmMsg::Execute {
+        contract_addr: config.usdclp_address.to_string(),
+        msg: to_json_binary(&Cw20ExecuteMsg::BurnFrom {
+            owner: sender.to_string(),
+            amount,
+        })?,
+        funds: vec![],
+    };
+
+    // Update the pending withdrawal amount and reduce the LP supply
+    state::PENDING_WITHDRAWALS.update(
```

Comment:
One problem with this approach: it gives priority to users based on lexicographic ordering of their addresses, not based on first-come-first-served. I'd rather go in the latter direction, even if it involves a bit more complexity, to avoid unfairness in funds handling.